From: Jeff Mahoney <jeffm@suse.com>
Subject: [PATCH 01/08] idr: add idr_replace method for replacing pointers
References: 163243
Patch-mainline: 2.6.18-rc1

 This patch adds an idr_replace() method to the IDR library for replacing
 a pointer already present in the IDR. Rather than do a remove/add pair, this
 function simply updates the pointer for an ID.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>
--
 include/linux/idr.h |    1 +
 lib/idr.c           |   37 +++++++++++++++++++++++++++++++++++++
 2 files changed, 38 insertions(+)

diff -ruNpX ../dontdiff linux-2.6.16-staging1/include/linux/idr.h linux-2.6.16-staging2/include/linux/idr.h
--- linux-2.6.16-staging1/include/linux/idr.h	2006-01-02 22:21:10.000000000 -0500
+++ linux-2.6.16-staging2/include/linux/idr.h	2006-04-13 16:18:14.000000000 -0400
@@ -78,6 +78,7 @@ void *idr_find(struct idr *idp, int id);
 int idr_pre_get(struct idr *idp, gfp_t gfp_mask);
 int idr_get_new(struct idr *idp, void *ptr, int *id);
 int idr_get_new_above(struct idr *idp, void *ptr, int starting_id, int *id);
+int idr_replace(struct idr *idp, void *ptr, int id);
 void idr_remove(struct idr *idp, int id);
 void idr_destroy(struct idr *idp);
 void idr_init(struct idr *idp);
diff -ruNpX ../dontdiff linux-2.6.16-staging1/lib/idr.c linux-2.6.16-staging2/lib/idr.c
--- linux-2.6.16-staging1/lib/idr.c	2006-01-02 22:21:10.000000000 -0500
+++ linux-2.6.16-staging2/lib/idr.c	2006-04-13 16:18:14.000000000 -0400
@@ -390,6 +390,43 @@ void *idr_find(struct idr *idp, int id)
 }
 EXPORT_SYMBOL(idr_find);
 
+/**
+ * idr_update - update pointer for given id
+ * @idp: idr handle
+ * @ptr: pointer you want associated with the ide
+ * @id: lookup key
+ *
+ * Update the pointer registered with the id.  A -ENOENT
+ * return indicates that @id is not valid.
+ *
+ * The caller must serialize vs idr_find(), idr_get_new(), and idr_remove().
+ */
+int idr_replace(struct idr *idp, void *ptr, int id)
+{
+	int n;
+	struct idr_layer *p;
+	int shift = (idp->layers - 1) * IDR_BITS;
+
+	n = idp->layers * IDR_BITS;
+	p = idp->top;
+
+	id &= MAX_ID_MASK;
+
+	while ((shift > 0) && p) {
+		n = (id >> shift) & IDR_MASK;
+		p = p->ary[n];
+		shift -= IDR_BITS;
+	}
+
+	n = id & IDR_MASK;
+	if (unlikely(p == NULL || !test_bit(n, &p->bitmap)))
+		return -ENOENT;
+
+	p->ary[n] = ptr;
+	return 0;
+}
+EXPORT_SYMBOL(idr_replace);
+
 static void idr_cache_ctor(void * idr_layer, kmem_cache_t *idr_layer_cache,
 		unsigned long flags)
 {
